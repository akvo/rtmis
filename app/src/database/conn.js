import { Asset } from 'expo-asset';
import * as FileSystem from 'expo-file-system';
import * as SQLite from 'expo-sqlite';

const openDatabase = () => {
  const db = SQLite.openDatabase('db.db');
  return db;
};

const init = openDatabase();

const tx = (db, query, params = []) => {
  return new Promise((resolve, reject) => {
    db.transaction(
      (transaction) => {
        if (Array.isArray(query)) {
          const promises = query.map((q) => {
            return new Promise((innerResolve, innerReject) => {
              transaction.executeSql(
                q,
                params,
                (_, resultSet) => {
                  innerResolve(resultSet);
                },
                (_, error) => {
                  innerReject(error);
                  return false; // Rollback the transaction
                },
              );
            });
          });

          Promise.all(promises)
            .then((results) => {
              resolve(results);
            })
            .catch((error) => {
              reject(error);
            });
        } else {
          transaction.executeSql(
            query,
            params,
            (_, resultSet) => {
              resolve(resultSet);
            },
            (_, error) => {
              reject(error);
              return false; // Rollback the transaction
            },
          );
        }
      },
      (error) => {
        reject(error);
      },
    );
  });
};

const openDBfile = async (databaseFile, databaseName) => {
  if (!(await FileSystem.getInfoAsync(FileSystem.documentDirectory + 'SQLite')).exists) {
    await FileSystem.makeDirectoryAsync(FileSystem.documentDirectory + 'SQLite');
  }
  await FileSystem.downloadAsync(
    Asset.fromModule(databaseFile).uri,
    FileSystem.documentDirectory + `SQLite/${databaseName}.db`,
  );
  return SQLite.openDatabase(`${databaseName}.db`);
};

export const conn = {
  file: (dbFile, dbName) => openDBfile(dbFile, dbName),
  init,
  tx,
};
