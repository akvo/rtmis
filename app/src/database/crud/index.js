export { default as crudSessions } from './crud-sessions';
export { default as crudForms } from './crud-forms';
export { default as crudUsers } from './crud-users';
export { default as crudConfig } from './crud-config';
export { default as crudDataPoints } from './crud-datapoints';
