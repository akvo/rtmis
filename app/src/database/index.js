export * from './conn';
export * from './query';
export * from './tables';
