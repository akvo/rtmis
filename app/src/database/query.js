const insert = (table, data = {}) => {
  const fields = Object.keys(data);
  const valuesString = fields
    .map((key) => (isNaN(data[key]) ? `'${data[key]}'` : data[key]))
    .join(', ');
  const fieldsString = fields.join(', ');
  return `INSERT INTO ${table}(${fieldsString}) VALUES (${valuesString});`;
};

const update = (table, where = {}, data = {}) => {
  const fieldString = Object.keys(data)
    .map((key) => (isNaN(data[key]) ? `${key} = '${data[key]}'` : `${key} = ${data[key]}`))
    .join(', ');
  const conditions = Object.keys(where).map((key) => `${key} = ?`);
  const conditionString = conditions.length ? `WHERE ${conditions.join(' AND ')}` : '';
  return `UPDATE ${table} SET ${fieldString} ${conditionString};`;
};

const read = (table, where = {}, nocase = false) => {
  const conditions = Object.keys(where).map((key) => {
    return where[key] === null ? `${key} IS NULL` : `${key} = ?`;
  });
  let conditionString = '';
  if (conditions.length) {
    conditionString = `WHERE ${conditions.join(' AND ')}`;
    if (nocase) {
      conditionString += ' COLLATE NOCASE';
    }
  }
  return `SELECT * FROM ${table} ${conditionString};`;
};

const clear = (tables = []) => {
  return tables.map((t) => `DELETE FROM ${t};`);
};

const drop = (table) => {
  return `DROP TABLE IF EXISTS ${table};`;
};

const count = (table) => {
  return `SELECT COUNT(*) AS count FROM ${table}`;
};

const initialQuery = (tableName, columns) => {
  const fields = Object.keys(columns).map((key) => `${key} ${columns[key]}`);
  const fieldsString = fields.join(', ');
  return `CREATE TABLE IF NOT EXISTS ${tableName}(${fieldsString});`;
};
export const query = {
  insert,
  update,
  read,
  clear,
  drop,
  count,
  initialQuery,
};
