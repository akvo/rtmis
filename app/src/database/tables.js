export const tables = [
  {
    name: 'users',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      name: 'TEXT',
      password: 'TEXT',
      active: 'TINYINT',
    },
  },
  {
    name: 'config',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      appVersion: 'VARCHAR(255) NOT NULL',
      authenticationCode: 'TEXT',
      serverURL: 'TEXT',
      syncInterval: 'REAL',
      syncWifiOnly: 'TINYINT',
      lang: 'VARCHAR(255) DEFAULT "en" NOT NULL',
    },
  },
  {
    name: 'forms',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      formId: 'INTEGER NOT NULL',
      version: 'VARCHAR(255)',
      latest: 'TINYINT',
      name: 'VARCHAR(255)',
      json: 'TEXT',
      createdAt: 'DATETIME',
    },
  },
  {
    name: 'datapoints',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      form: 'INTEGER NOT NULL',
      user: 'INTEGER NOT NULL',
      name: 'VARCHAR(255)',
      geo: 'VARCHAR(255)',
      submitted: 'TINYINT',
      duration: 'REAL',
      createdAt: 'DATETIME',
      submittedAt: 'DATETIME',
      syncedAt: 'DATETIME',
      json: 'TEXT',
    },
  },
  {
    name: 'sessions',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      token: 'TEXT',
      passcode: 'TEXT',
    },
  },
];
