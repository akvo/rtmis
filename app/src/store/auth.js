import { Store } from 'pullstate';
/**
 * Server config
 */
const AuthState = new Store({
  authenticationCode: '',
  useAuthenticationCode: false, // using code for authentication
  username: '',
  password: '',
  token: null,
});

export default AuthState;
