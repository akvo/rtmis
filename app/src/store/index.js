export { default as AuthState } from './auth';
export { default as BuildParamsState } from './buildParams';
export { default as FormState } from './forms';
export { default as UIState } from './ui';
export { default as UserState } from './users';
