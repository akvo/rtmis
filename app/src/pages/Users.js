import React, { useState, useEffect, useCallback } from 'react';
import { ScrollView, BackHandler, Platform, ToastAndroid } from 'react-native';
import { Button, ListItem, Skeleton } from '@rneui/themed';
import Icon from 'react-native-vector-icons/Ionicons';

import { BaseLayout } from '../components';
import { conn, query } from '../database';
import { UserState, UIState } from '../store';
import { i18n } from '../lib';

const db = conn.init;

const Users = ({ navigation, route }) => {
  const [loading, setLoading] = useState(true);
  const [users, setUsers] = useState([]);
  const currUserID = UserState.useState((s) => s.id);
  const activeLang = UIState.useState((s) => s.lang);
  const trans = i18n.text(activeLang);

  const goToCreate = () => {
    navigation.navigate('AddUser');
  };

  const goToHome = () => {
    navigation.navigate('Home');
  };

  const loadUsers = useCallback(async () => {
    const selectQuery = query.read('users');
    const { rows } = await conn.tx(db, selectQuery);
    setUsers(rows._array);
    setLoading(false);
  }, []);

  const handleSelectUser = async (id, name) => {
    const currUserQuery = query.update('users', { id: currUserID }, { active: 0 });
    await conn.tx(db, currUserQuery, [currUserID]);

    const thisUserQuery = query.update('users', { id }, { active: 1 });
    await conn.tx(db, thisUserQuery, [id]);

    UserState.update((s) => {
      s.id = id;
      s.name = name;
    });
    await loadUsers();

    if (Platform.OS === 'android') {
      ToastAndroid.show(`${trans.usersSwitchTo}${name}`, ToastAndroid.SHORT);
    }
  };

  useEffect(() => {
    if (loading) {
      loadUsers();
    }
    if (!loading && route?.params?.added) {
      const newUser = route.params.added;
      const findNew = users.find((u) => {
        return u.id === newUser?.id;
      });
      if (!findNew) {
        setLoading(true);
      }
    }
  }, [loading, route, loadUsers]);

  useEffect(() => {
    const handleBackPress = () => {
      goToHome(); // Change the destination force to 'Home'
      return true; // Return true to prevent default back behavior
    };
    const backHandler = BackHandler.addEventListener('hardwareBackPress', handleBackPress);
    return () => {
      backHandler.remove(); // Cleanup the event listener on component unmount
    };
  }, []);

  return (
    <BaseLayout
      title={trans.usersPageTitle}
      leftComponent={
        <Button type="clear" onPress={goToHome} testID="arrow-back-button">
          <Icon name="arrow-back" size={18} />
        </Button>
      }
      rightComponent={
        <Button type="clear" testID="button-add-user" onPress={goToCreate}>
          <Icon name="add" size={18} />
        </Button>
      }
    >
      <ScrollView>
        {loading && <Skeleton animation="wave" testID="loading-users" />}
        {users.map((user, index) => {
          return (
            <ListItem.Swipeable
              key={index}
              onPress={async () => await handleSelectUser(user.id, user.name)}
              rightContent={(reset) => (
                <Button
                  title={trans.buttonDelete}
                  onPress={() => reset()}
                  icon={{ name: 'delete', color: 'white' }}
                  buttonStyle={{ minHeight: '100%', backgroundColor: 'red' }}
                />
              )}
              testID={`list-item-user-${user.id}`}
              bottomDivider
            >
              <ListItem.Content>
                <ListItem.Title testID={`title-username-${user.id}`}>{user.name}</ListItem.Title>
              </ListItem.Content>
              {user.active === 1 && (
                <Icon name="checkmark" size={18} testID={`icon-checkmark-${user.id}`} />
              )}
            </ListItem.Swipeable>
          );
        })}
      </ScrollView>
    </BaseLayout>
  );
};

export default Users;
