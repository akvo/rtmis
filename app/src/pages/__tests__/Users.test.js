import React, { useState } from 'react';
import { create } from 'react-test-renderer';
import { render, renderHook, fireEvent, act, waitFor } from '@testing-library/react-native';
import { useNavigation } from '@react-navigation/native';
import mockBackHandler from 'react-native/Libraries/Utilities/__mocks__/BackHandler.js';

import Users from '../Users';
import { UserState, UIState } from '../../store';
import { conn, query } from '../../database';

jest.mock('expo-sqlite');

jest.mock('react', () => ({
  ...jest.requireActual('react'),
}));

jest.mock('react-native/Libraries/Utilities/BackHandler', () => mockBackHandler);

db = conn.init;

describe('UsersPage', () => {
  beforeAll(() => {
    UIState.update((s) => {
      s.lang = 'en';
    });
  });
  beforeEach(() => {
    const usersData = [
      {
        id: 1,
        name: 'John',
        active: 1,
      },
    ];
    const mockSelectSql = jest.fn((query, params, successCallback) => {
      successCallback(null, { rows: { length: usersData.length, _array: usersData } });
    });

    db.transaction.mockImplementation((transactionFunction) => {
      transactionFunction({
        executeSql: mockSelectSql,
      });
    });
  });

  it('should match with snapshot', async () => {
    const mockNavigation = useNavigation();
    const { toJSON } = render(<Users navigation={mockNavigation} />);
    const { result } = renderHook(() => useState([]));
    const { result: resLoading } = renderHook(() => useState(true));
    const [users, setUsers] = result.current;
    const [loading, setLoading] = resLoading.current;

    await act(async () => {
      const selectQuery = query.read('users');
      const rows = await conn.tx(db, selectQuery);
      setUsers(rows._array);
      setLoading(false);
    });

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('should show empty page when users table has no data', () => {
    const mockNavigation = useNavigation();
    const mockSelectSql = jest.fn((query, params, successCallback) => {
      successCallback(null, { rows: { length: 0, _array: [] } });
    });
    db.transaction.mockImplementation((transactionFunction) => {
      transactionFunction({
        executeSql: mockSelectSql,
      });
    });

    const { queryByTestId } = render(<Users navigation={mockNavigation} />);

    const userItemEl = queryByTestId('list-item-user-1');
    expect(userItemEl).toBeNull();
  });

  it('should render users list when the table has data', async () => {
    const mockNavigation = useNavigation();
    const usersData = [
      {
        id: 1,
        name: 'John Doe',
        active: 1,
      },
      {
        id: 2,
        name: 'Miles Morales',
        active: 0,
      },
    ];
    const mockSelectSql = jest.fn((query, params, successCallback) => {
      successCallback(null, { rows: { length: usersData.length, _array: usersData } });
    });

    db.transaction.mockImplementation((transactionFunction) => {
      transactionFunction({
        executeSql: mockSelectSql,
      });
    });
    const { getByText } = render(<Users navigation={mockNavigation} />);
    const mockLoadUsers = jest.fn();
    act(() => {
      mockLoadUsers();
    });

    const { result } = renderHook(() => useState([]));
    const { result: resLoading } = renderHook(() => useState(true));
    const [users, setUsers] = result.current;
    const [loading, setLoading] = resLoading.current;

    await act(async () => {
      const selectQuery = query.read('users');
      const rows = await conn.tx(db, selectQuery);
      setUsers(rows._array);
      setLoading(false);
    });

    await waitFor(() => {
      expect(mockLoadUsers).toHaveBeenCalledTimes(1);
      expect(resLoading.current[0]).toBeFalsy();
      expect(getByText('John Doe')).toBeDefined();
      expect(getByText('Miles Morales')).toBeDefined();
    });
  });

  it('should show check mark for active user', async () => {
    const mockNavigation = useNavigation();
    const { getByTestId } = render(<Users navigation={mockNavigation} />);
    const { result } = renderHook(() => useState([]));
    const { result: resLoading } = renderHook(() => useState(true));
    const [users, setUsers] = result.current;
    const [loading, setLoading] = resLoading.current;

    await act(async () => {
      const selectQuery = query.read('users');
      const rows = await conn.tx(db, selectQuery);
      setUsers(rows._array);
      setLoading(false);
    });

    await waitFor(() => {
      const iconCheckMark = getByTestId('icon-checkmark-1');
      expect(iconCheckMark).toBeDefined();
    });
  });

  it('should be able to switch to another user', async () => {
    const mockNavigation = useNavigation();
    const usersData = [
      {
        id: 1,
        name: 'John Doe',
        active: 1,
      },
      {
        id: 2,
        name: 'Miles Morales',
        active: 0,
      },
    ];
    const mockSelectSql = jest.fn((query, params, successCallback) => {
      successCallback(null, { rows: { length: usersData.length, _array: usersData } });
    });

    db.transaction.mockImplementation((transactionFunction) => {
      transactionFunction({
        executeSql: mockSelectSql,
      });
    });

    const { getByTestId, rerender } = render(<Users navigation={mockNavigation} />);

    const { result } = renderHook(() => useState([]));
    const { result: resLoading } = renderHook(() => useState(true));
    const [users, setUsers] = result.current;
    const [loading, setLoading] = resLoading.current;

    await act(async () => {
      const selectQuery = query.read('users');
      const rows = await conn.tx(db, selectQuery);
      setUsers(rows._array);
      setLoading(false);
    });

    rerender(<Users navigation={mockNavigation} />);

    const selectedUser = getByTestId('list-item-user-2');

    expect(selectedUser).toBeDefined();
    fireEvent.press(selectedUser);

    const updatedUserData = [
      {
        id: 1,
        name: 'John Doe',
        active: 0,
      },
      {
        id: 2,
        name: 'Miles Morales',
        active: 1,
      },
    ];
    const mockUpdatedSql = jest.fn((query, params, successCallback) => {
      successCallback(null, { rows: { length: updatedUserData.length, _array: updatedUserData } });
    });

    db.transaction.mockImplementation((transactionFunction) => {
      transactionFunction({
        executeSql: mockUpdatedSql,
      });
    });

    act(() => {
      UserState.update((s) => {
        s.id = 2;
        s.name = 'Miles Morales';
      });
    });

    await waitFor(() => {
      const { result } = renderHook(() => UserState.useState((s) => s.name));
      const selectedUserChecked = getByTestId('icon-checkmark-2');
      expect(selectedUserChecked).toBeDefined();
      expect(result.current).toBe('Miles Morales');
    });
  });

  it('should go to add user when plus icon clicked', () => {
    const mockNavigation = useNavigation();
    const { getByTestId, rerender } = render(<Users navigation={mockNavigation} />);
    const addUserButton = getByTestId('button-add-user');
    expect(addUserButton).toBeDefined();
    fireEvent.press(addUserButton);

    expect(mockNavigation.navigate).toHaveBeenCalledWith('AddUser');
  });

  it('should redirect to Homepage when hardware back button pressed', async () => {
    const mockNavigation = useNavigation();
    render(<Users navigation={mockNavigation} />);

    act(() => {
      mockBackHandler.mockPressBack();
    });

    await waitFor(() => {
      expect(mockNavigation.navigate).toHaveBeenCalledWith('Home');
    });
  });

  it('should reload page after new user added', async () => {
    const mockNavigation = useNavigation();
    const mockRoute = {
      params: {
        added: {
          id: 3,
          name: 'Barry Allen',
        },
      },
    };
    const { getByText } = render(<Users navigation={mockNavigation} route={mockRoute} />);
    const { result } = renderHook(() => useState([]));
    const { result: resLoading } = renderHook(() => useState(true));
    const [users, setUsers] = result.current;
    const [loading, setLoading] = resLoading.current;

    const mockLoadUsers = jest.fn();
    act(() => {
      mockLoadUsers();
    });
    const updatedUserData = [
      {
        id: 1,
        name: 'John',
        active: 1,
      },
      {
        id: 3,
        name: 'Barry Allen',
        active: 0,
      },
    ];
    const mockUpdatedSql = jest.fn((query, params, successCallback) => {
      successCallback(null, { rows: { length: updatedUserData.length, _array: updatedUserData } });
    });

    db.transaction.mockImplementation((transactionFunction) => {
      transactionFunction({
        executeSql: mockUpdatedSql,
      });
    });
    act(() => {
      mockLoadUsers();
    });

    await waitFor(() => {
      expect(mockLoadUsers).toHaveBeenCalledTimes(2);
      expect(getByText('John')).toBeDefined();
      expect(getByText('Barry Allen')).toBeDefined();
    });
  });

  it('should be translated when UIState.lang changed', async () => {
    const mockNavigation = useNavigation();
    const { getByText } = render(<Users navigation={mockNavigation} />);

    const enTitle = getByText('Users');
    expect(enTitle).toBeDefined();

    act(() => {
      UIState.update((s) => {
        s.lang = 'fr';
      });
    });

    await waitFor(() => {
      const frTitle = getByText('Utilisateurs');
      expect(frTitle).toBeDefined();
    });
  });
});
