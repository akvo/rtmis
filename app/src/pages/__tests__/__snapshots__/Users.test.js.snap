// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersPage should match with snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9fafb",
      "flex": 1,
      "flexDirection": "column",
      "flexWrap": "wrap",
    }
  }
  testID="stack-container"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f3f4f6",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "minheight": 78,
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
    testID="base-layout-page-title"
  >
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 2,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="arrow-back-button"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 2,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                  "paddingHorizontal": 10,
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 18,
                    },
                    undefined,
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontSize": 18,
              "paddingTop": 4,
              "textAlign": "center",
            }
          }
          testID="page-title"
        >
          Users
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 2,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="button-add-user"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 2,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                  "paddingHorizontal": 10,
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 18,
                    },
                    undefined,
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RNCSafeAreaView>
  </View>
  <RCTScrollView
    style={
      {
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          {
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            {
              "bottom": 0,
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              [
                {
                  "width": 250,
                  "zIndex": 1,
                },
                undefined,
              ]
            }
          />
          <View
            style={
              {
                "flex": 0,
              }
            }
          />
          <View
            style={
              [
                {
                  "width": 250,
                  "zIndex": 1,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 2,
                  },
                  undefined,
                  false,
                ]
              }
              testID="RNE_BUTTON_WRAPPER"
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": false,
                    "checked": undefined,
                    "disabled": false,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "red",
                      "borderColor": "#2089dc",
                      "borderRadius": 2,
                      "borderWidth": 0,
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "minHeight": "100%",
                      "padding": 8,
                      "paddingHorizontal": 10,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "marginHorizontal": 5,
                        "overflow": "hidden",
                      }
                    }
                    testID="RNE__ICON__CONTAINER"
                  >
                    <View
                      accessibilityRole="button"
                      android_ripple={
                        {
                          "borderless": false,
                          "color": "rgba(0, 0, 0, 0.3)",
                          "radius": -5,
                        }
                      }
                      disabled={false}
                      testID="RNE__ICON__CONTAINER_ACTION"
                    >
                      <View
                        style={
                          {
                            "alignItems": "center",
                            "backgroundColor": "transparent",
                            "justifyContent": "center",
                          }
                        }
                        testID="RNE__ICON"
                      >
                        <Text
                          allowFontScaling={false}
                          selectable={false}
                          style={
                            [
                              {
                                "color": "white",
                                "fontSize": 24,
                              },
                              {
                                "backgroundColor": "transparent",
                              },
                              {
                                "fontFamily": "material",
                                "fontStyle": "normal",
                                "fontWeight": "normal",
                              },
                              {},
                            ]
                          }
                          testID="RNE__ICON__Component"
                        >
                          
                        </Text>
                      </View>
                    </View>
                  </View>
                  <Text
                    style={
                      {
                        "color": "white",
                        "fontSize": 18,
                        "paddingVertical": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    Delete
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          onMoveShouldSetResponder={[Function]}
          onMoveShouldSetResponderCapture={[Function]}
          onResponderEnd={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderReject={[Function]}
          onResponderRelease={[Function]}
          onResponderStart={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onStartShouldSetResponderCapture={[Function]}
          style={
            {
              "transform": [
                {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="list-item-user-1"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#ffffff",
                  "borderBottomWidth": 0.5,
                  "borderColor": "#bcbbc1",
                  "flexDirection": "row",
                  "padding": 14,
                }
              }
              testID="RNE__LISTITEM__padView"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "flex-start",
                      "flex": 1,
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
                theme={
                  {
                    "colors": {
                      "background": "#ffffff",
                      "black": "#242424",
                      "disabled": "hsl(208, 8%, 90%)",
                      "divider": "#bcbbc1",
                      "error": "#ff190c",
                      "grey0": "#393e42",
                      "grey1": "#43484d",
                      "grey2": "#5e6977",
                      "grey3": "#86939e",
                      "grey4": "#bdc6cf",
                      "grey5": "#e1e8ee",
                      "greyOutline": "#bbb",
                      "platform": {
                        "android": {
                          "error": "#f44336",
                          "grey": "rgba(0, 0, 0, 0.54)",
                          "primary": "#2196f3",
                          "searchBg": "#dcdce1",
                          "secondary": "#9C27B0",
                          "success": "#4caf50",
                          "warning": "#ffeb3b",
                        },
                        "default": {
                          "error": "#ff3b30",
                          "grey": "#7d7d7d",
                          "primary": "#007aff",
                          "searchBg": "#dcdce1",
                          "secondary": "#5856d6",
                          "success": "#4cd964",
                          "warning": "#ffcc00",
                        },
                        "ios": {
                          "error": "#ff3b30",
                          "grey": "#7d7d7d",
                          "primary": "#007aff",
                          "searchBg": "#dcdce1",
                          "secondary": "#5856d6",
                          "success": "#4cd964",
                          "warning": "#ffcc00",
                        },
                        "web": {
                          "error": "#ff190c",
                          "grey": "#393e42",
                          "primary": "#2089dc",
                          "searchBg": "#303337",
                          "secondary": "#ca71eb",
                          "success": "#52c41a",
                          "warning": "#faad14",
                        },
                      },
                      "primary": "#2089dc",
                      "searchBg": "#303337",
                      "secondary": "#ad1457",
                      "success": "#52c41a",
                      "warning": "#faad14",
                      "white": "#ffffff",
                    },
                    "spacing": {
                      "lg": 12,
                      "md": 8,
                      "sm": 4,
                      "xl": 24,
                      "xs": 2,
                    },
                  }
                }
              >
                <Text
                  accessibilityRole="text"
                  style={
                    {
                      "backgroundColor": "transparent",
                      "color": "#242424",
                      "fontSize": 17,
                    }
                  }
                  testID="title-username-1"
                >
                  John
                </Text>
              </View>
              <View
                style={
                  {
                    "paddingLeft": 16,
                  }
                }
              />
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 18,
                    },
                    undefined,
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="icon-checkmark-1"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
